xdist/scheduler/loadsuite.py

```diff
@@ -0,0 +1,413 @@
+
+        Called by the hooks:
+
+        - ``DSession.slave_slavefinished``.
+        - ``DSession.slave_errordown``.
+
+        Return the item being executed while the node crashed or None if the
+        node has no more pending items.
+        """
+        workload = self.assigned_work.pop(node)
+        if not self._pending_of(workload):
+            return None
+
+        # The node crashed, identify test that crashed
+        for work_unit in workload.values():
+            for nodeid, completed in work_unit.items():
+                if not completed:
+                    crashitem = nodeid
+                    break
+            else:
+                continue
+            break
+        else:
+            raise RuntimeError(
+                'Unable to identify crashitem on a workload with '
+                'pending items'
+            )
+
+        # Made uncompleted work unit available again
+        self.workqueue.update(workload)
+
+        for node in self.assigned_work:
+            self._reschedule(node)
+
+        return crashitem
+
+    def add_node_collection(self, node, collection):
+        """
+        Add the collected test items from a node.
+
+        The collection is stored in the ``.registered_collections`` dictionary.
+
+        Called by the hook:
+
+        - ``DSession.slave_collectionfinish``.
+        """
+        assert node in self.assigned_work
+
+        if not self.collection_is_completed:
+            self.registered_collections[node] = list(collection)
+            return
+
+        # A new node has been added later, perhaps an original one died.
+        # .schedule() should have been called by now
+        assert self.collection
+        if collection != self.collection:
+
+            other_node = next(iter(self.registered_collections.keys()))
+
+            msg = report_collection_diff(
+                self.collection,
+                collection,
+                other_node.gateway.id,
+                node.gateway.id
+            )
+            self.log(msg)
+
+    def mark_test_complete(self, node, item_index, duration=0):
+        """
+        Mark test item as completed by node.
+
+        Called by the hook:
+
+        - ``DSession.slave_testreport``.
+        """
+        nodeid = self.registered_collections[node][item_index]
+        suite = nodeid.split('::', 1)[0]
+
+        self.assigned_work[node][suite][nodeid] = True
+        self._reschedule(node)
+
+    def _asign_work_unit(self, node):
+        """
+        Assign a work unit to a node.
+        """
+        assert self.workqueue
+
+        # Grab a unit of work
+        suite, work_unit = self.workqueue.popitem(last=False)
+
+        # Keep track of the assigned work
+        assigned_to_node = self.assigned_work.setdefault(
+            node, default=OrderedDict()
+        )
+        assigned_to_node[suite] = work_unit
+
+        # Ask the node to execute the workload
+        worker_collection = self.registered_collections[node]
```

Comment:
Sorry if I'm missing something, but wouldn't be simpler to keep a single dict mapping nodeid to its completed status instead of each node has its own separate data structure? We already guarantee that all workers have collected the exact  same items, so we can use that assumption to drop `registered_collections`. Less data structures to keep things in sync the better IMHO.

I think you can change `self.collection` into a dict of suite -> list of item ids.